## Programa Ingenias+ Data Science

## Trabajando con Pandas

1) Importa las librerias requeridas: `Pandas` y `NumPy`.  

Una buena practica a la hora de escribir codigo es de comentar siempre que estoy haciendo (usando `#`) en la linea anterior. Por mas obvio que parezca lo que voy a hacer, puede ser que tenga que compartir el documento con otras personas y no sepan que quise hacer en cada linea. O que yo mismo lea el codigo tiempo despues y no recuerde porque lo hice.

In [ ]:
#Importar Pandas
import pandas as pd

#Importar Numpy
import numpy as np


2) Lee el archivo csv `life_expectancy_data` usando `pandas`. Guardalo en una variable llamada `expvida`.

In [13]:
# Leo los datos en un DataFrame
import pandas as pd
df=pd.read_csv('life_expectancy_data.csv',delimiter=',')
expvida = "life_expectancy_data.csv"


3) ¿Que tipo de estructura de datos contiene la variable `expvida`?

In [17]:
# Obtengo el tipo de estructura de datos
expvida = "life_expectancy_data.csv"
tipo_expvida = type(expvida)
print(tipo_expvida)



<class 'str'>


4) ¿Cuantas filas y columnas tiene `expvida`?

In [24]:
# Uso .shape para obtener cantidad de filas y columnas
datos = pd.read_csv(expvida)

# Obtener la forma (shape) del DataFrame
filas, columnas = datos.shape
print("FILAS:", filas)
print("COLUMNAS:", columnas)



FILAS: 2938
COLUMNAS: 22


**Numero de filas**: 2938

**Numero de columnas**: 22

5) ¿Cual es el nombre de las columnas contenidas en `expvida`?

In [25]:
# Obtengo el nombre de las columnas

datos = pd.read_csv(expvida)

# Obtener los nombres de las columnas
nombres_columnas = datos.columns
print(nombres_columnas)


Index(['Country', 'Year', 'Status', 'Life expectancy ', 'Adult Mortality',
       'infant deaths', 'Alcohol', 'percentage expenditure', 'Hepatitis B',
       'Measles ', ' BMI ', 'under-five deaths ', 'Polio', 'Total expenditure',
       'Diphtheria ', ' HIV/AIDS', 'GDP', 'Population',
       ' thinness  1-19 years', ' thinness 5-9 years',
       'Income composition of resources', 'Schooling'],
      dtype='object')


6) Inspecciona las primeras 10 filas de `expvida`

In [27]:

# Inspeccionar las primeras filas de expvida
import pandas as pd
datos=pd.read_csv(expvida)
# Obtener las primeras diez filas como ejemplo
primeras_filas= datos.head(10)
print("Filas:")
print(primeras_filas)



Filas:
       Country  Year      Status  Life expectancy   Adult Mortality  \
0  Afghanistan  2015  Developing              65.0            263.0   
1  Afghanistan  2014  Developing              59.9            271.0   
2  Afghanistan  2013  Developing              59.9            268.0   
3  Afghanistan  2012  Developing              59.5            272.0   
4  Afghanistan  2011  Developing              59.2            275.0   
5  Afghanistan  2010  Developing              58.8            279.0   
6  Afghanistan  2009  Developing              58.6            281.0   
7  Afghanistan  2008  Developing              58.1            287.0   
8  Afghanistan  2007  Developing              57.5            295.0   
9  Afghanistan  2006  Developing              57.3            295.0   

   infant deaths  Alcohol  percentage expenditure  Hepatitis B  Measles   ...  \
0             62     0.01               71.279624         65.0      1154  ...   
1             64     0.01               73.523582

7) Y ahora las 10 ultimas.

In [28]:
# Inspecciono las ultimas filas
import pandas as pd

datos=pd.read_csv(expvida)
ultimas_filas=datos.tail(10)

print("ultimas:")
print(ultimas_filas)

ultimas:
       Country  Year      Status  Life expectancy   Adult Mortality  \
2928  Zimbabwe  2009  Developing              50.0            587.0   
2929  Zimbabwe  2008  Developing              48.2            632.0   
2930  Zimbabwe  2007  Developing              46.6             67.0   
2931  Zimbabwe  2006  Developing              45.4              7.0   
2932  Zimbabwe  2005  Developing              44.6            717.0   
2933  Zimbabwe  2004  Developing              44.3            723.0   
2934  Zimbabwe  2003  Developing              44.5            715.0   
2935  Zimbabwe  2002  Developing              44.8             73.0   
2936  Zimbabwe  2001  Developing              45.3            686.0   
2937  Zimbabwe  2000  Developing              46.0            665.0   

      infant deaths  Alcohol  percentage expenditure  Hepatitis B  Measles   \
2928             30     4.64                1.040021         73.0       853   
2929             30     3.56               20.84342

8) ¿Que tipos de datos contiene cada una de las columnas de `expvida`?

In [30]:
# Tipo de datos en cada columna
import pandas as pd
datos=pd.read_csv(expvida)
tipos_de_datos=datos.dtypes

print("Tipos de datos:")
print(tipos_de_datos)

Tipos de datos:
Country                             object
Year                                 int64
Status                              object
Life expectancy                    float64
Adult Mortality                    float64
infant deaths                        int64
Alcohol                            float64
percentage expenditure             float64
Hepatitis B                        float64
Measles                              int64
 BMI                               float64
under-five deaths                    int64
Polio                              float64
Total expenditure                  float64
Diphtheria                         float64
 HIV/AIDS                          float64
GDP                                float64
Population                         float64
 thinness  1-19 years              float64
 thinness 5-9 years                float64
Income composition of resources    float64
Schooling                          float64
dtype: object


9) ¿Hay valores faltantes en alguna de las columnas de `expvida`?

In [31]:
# Obtengo la cantidad de datos faltantes en cada columna
import pandas as pd
datos=pd.read_csv(expvida)

valores_nulos=datos.isnull().any()
print("Valores nulos:")
print(valores_nulos)

Valores nulos:
Country                            False
Year                               False
Status                             False
Life expectancy                     True
Adult Mortality                     True
infant deaths                      False
Alcohol                             True
percentage expenditure             False
Hepatitis B                         True
Measles                            False
 BMI                                True
under-five deaths                  False
Polio                               True
Total expenditure                   True
Diphtheria                          True
 HIV/AIDS                          False
GDP                                 True
Population                          True
 thinness  1-19 years               True
 thinness 5-9 years                 True
Income composition of resources     True
Schooling                           True
dtype: bool


10) Selecciona el maximo de datos faltantes. Calcula que porcentaje de los datos representan los datos faltantes.

In [32]:
# Obtengo el maximo valor de datos faltantes y lo guardo en la variable masnan
import pandas as pd
datos=pd.read_csv(expvida)
masnan=datos.isnull().sum().max()
print("Maximo valor:", masnan)

Maximo valor: 652


In [33]:
# Obtengo la cantidad de filas total y divido el maximo valor de datos faltantes por este numero
# y lo multiplico por 100
import pandas as pd

# Cargar el archivo CSV en un DataFrame
datos = pd.read_csv(expvida)

# Obtener la cantidad total de filas
total_filas = len(datos)

# Obtener el máximo valor de datos faltantes en todas las columnas
masnan = datos.isnull().sum().max()

# Calcular el porcentaje de datos faltantes
porcentaje_faltantes = (masnan / total_filas) * 100

print("Porcentaje de datos faltantes en relación con el total de filas:", porcentaje_faltantes)


Porcentaje de datos faltantes en relación con el total de filas: 22.19196732471069


In [34]:
# Imprimo el maximo porcentaje de datos faltantes. Uso f-strings.
import pandas as pd
datos = pd.read_csv(expvida)
total_filas = len(datos)
masnan = datos.isnull().sum().max()

# Calcular el porcentaje de datos faltantes
porcentaje_faltantes = (masnan / total_filas) * 100

# Imprimir el máximo porcentaje de datos faltantes utilizando f-strings
print(f"El máximo porcentaje de datos faltantes es: {porcentaje_faltantes:.2f}%")


El máximo porcentaje de datos faltantes es: 22.19%


- str.format()
- Template Class
- f-strings

#### Ahora vamos a realizar algunos calculos y manipulaciones sobre las columnas.

Una función útil para este fin es: `.rename()`. Y podemos utilizarla de la siguiente manera: Si `df` es mi DataFrame y quiero renombrar la columna `columna` como `columna1`, y la columna `columna2` como `columna3`, utilizo la siguiente sintaxis:
  
`df.rename(columns={"columna": "columna1", "columna2": "columna3"}, inplace=True)`

11) Renombra la columna 'Life expectancy ' como 'life_expectancy', ' BMI ' como 'bmi' y ''Measles ' como 'measles'. No olvides de usar `inplace` para modificar el DataFrame original. Luego chequea que el cambio se haya hecho correctamente.

In [37]:
# Renombro columnas del DataFrame
import pandas as pd
datos=pd.read_csv(expvida)
datos.rename(columns={'Life expectancy': 'life_expectancy','BMI':'bmi','Measles':'measles'},inplace=True)


In [38]:
# Chequeo que se hayan reemplazado correctamente
print("Columnas actualizadas:")
print(datos.columns)

Columnas actualizadas:
Index(['Country', 'Year', 'Status', 'Life expectancy ', 'Adult Mortality',
       'infant deaths', 'Alcohol', 'percentage expenditure', 'Hepatitis B',
       'Measles ', ' BMI ', 'under-five deaths ', 'Polio', 'Total expenditure',
       'Diphtheria ', ' HIV/AIDS', 'GDP', 'Population',
       ' thinness  1-19 years', ' thinness 5-9 years',
       'Income composition of resources', 'Schooling'],
      dtype='object')


12) ¿Cual es el promedio de expectativa de vida (Life expectancy) en general?

In [40]:
# Obtengo el promedio de la columna
import pandas as pd
datos = pd.read_csv(expvida)

# Calcular el promedio de expectativa de vida
promedio_expectativa_vida = datos['Life expectancy '].mean()

print("Promedio de expectativa de vida en general:", promedio_expectativa_vida)


Promedio de expectativa de vida en general: 69.22493169398908


13) ¿Cual es el promedio de expectativa de vida (Life expectancy) por país?

In [41]:
# Obtengo el promedio por pais
import pandas as pd
datos = pd.read_csv(expvida)

# Calcular el promedio de expectativa de vida por país
promedio_por_pais = datos.groupby('Country')['Life expectancy '].mean()

print("Promedio de expectativa de vida por país:")
print(promedio_por_pais)


Promedio de expectativa de vida por país:
Country
Afghanistan                           58.19375
Albania                               75.15625
Algeria                               73.61875
Angola                                49.01875
Antigua and Barbuda                   75.05625
                                        ...   
Venezuela (Bolivarian Republic of)    73.38750
Viet Nam                              74.77500
Yemen                                 63.86250
Zambia                                53.90625
Zimbabwe                              50.48750
Name: Life expectancy , Length: 193, dtype: float64


14) ¿Cual es la expectativa de vida mas alta y mas baja? Usa la funcion de pandas y la funcion de NumPy y compara el resultado. ¿Que paises tienen esas expectativas de vida?

In [47]:
import pandas as pd
import numpy as np

# Cargar el archivo CSV en un DataFrame
datos = pd.read_csv(expvida)

# Usando NumPy para encontrar la expectativa de vida más alta y más baja
max_numpy = np.max(datos['Life expectancy '])
min_numpy = np.min(datos['Life expectancy '])

# Obtener los países con la expectativa de vida más alta y más baja
pais_max_numpy = datos.loc[datos['Life expectancy '] == max_numpy, 'Country'].iloc[0]
pais_min_numpy = datos.loc[datos['Life expectancy '] == min_numpy, 'Country'].iloc[0]

print("Expectativa de vida más alta (NumPy):", max_numpy, "en", pais_max_numpy)
print("Expectativa de vida más baja (NumPy):", min_numpy, "en", pais_min_numpy)



Expectativa de vida más alta (NumPy): 89.0 en Belgium
Expectativa de vida más baja (NumPy): 36.3 en Haiti


In [ ]:
# Maxima expectativa de vida usando pandas


In [48]:
# Minima expectativa de vida usando numpy
pais_min_numpy = datos.loc[datos['Life expectancy '] == min_numpy, 'Country'].iloc[0]


In [ ]:
# Minima expectativa de vida usando pandas


In [ ]:
# Muestro los resultados maximos


In [ ]:
# Muestro los resultados minimos


In [ ]:
# Filtro las columnas para mostrar solo las filas que tengan valor
# igual al maximo


In [ ]:
# Filtro las columnas para mostrar solo las filas que tengan valor
# igual al minimo


15) ¿Cuantas categorias encontramos en la columna "Status"? _**Pista**_: Lee acerca de la función [unique()](http://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.Series.unique.html?highlight=unique#pandas.Series.unique)

In [52]:
# Obtengo los valores unicos de la columna status
import pandas as pd

# Cargar el archivo CSV en un DataFrame
datos = pd.read_csv(expvida)

# Obtener los valores únicos en la columna "Status"
categorias_status = datos['Status'].unique()

# Contar la cantidad de categorías
cantidad_categorias = len(categorias_status)

print("Cantidad de categorías en la columna 'Status':", cantidad_categorias)


Cantidad de categorías en la columna 'Status': 2


16) ¿Cuantos paises se ubican en cada una de esas categorias de "Status"? _**Pista**_: Lee acerca de la función [value_counts()](http://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.Series.value_counts.html)

In [51]:
# Cuento cuantos valores hay de cada categoria
import pandas as pd

# Cargar el archivo CSV en un DataFrame
datos = pd.read_csv(expvida)

# Contar la cantidad de países en cada categoría de "Status"
paises_por_status = datos['Status'].value_counts()

print("Cantidad de países en cada categoría de Status:")
print(paises_por_status)


Cantidad de países en cada categoría de Status:
Status
Developing    2426
Developed      512
Name: count, dtype: int64


18) ¿Cuantos paises presentaron algun caso de Sarampion (**Measles**) en el 2015? **_Pista_**: Lee acerca de la función [.count()](http://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.count.html?highlight=count#pandas.DataFrame.count)

In [50]:
# Cuento cuantas filas tienen valores distintos de cero
import pandas as pd

# Cargar el archivo CSV en un DataFrame
datos = pd.read_csv(expvida)

# Filtrar los datos para el año 2015
datos_2015 = datos[datos['Year'] == 2015]

# Contar cuántos países presentaron casos de sarampión en 2015
paises_con_sarampion_2015 = datos_2015['Measles '].count()

print("Cantidad de países con casos de sarampión en 2015:", paises_con_sarampion_2015)


Cantidad de países con casos de sarampión en 2015: 183
